### PR TITLE
feat: implement halo2 common primitives

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This PR adds common structures that used by `halo2-common` to `main`.

### CODE REVIEW

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Comment by TomTaehoonKim** [comment]
@chokobole Could you fix typo in the commit description in [1b140f1](https://github.com/kroma-network/tachyon/pull/106/commits/1b140f10a6e16ecd01af6fb9236501ed3381a2fc)? There is an unnecessary "i" in i[VirtualCell].

**Comment by chokobole** [comment]
@TomTaehoonKim I fixed it.

> @chokobole Could you fix typo in the commit description in [1b140f1](https://github.com/kroma-network/tachyon/pull/106/commits/1b140f10a6e16ecd01af6fb9236501ed3381a2fc)? There is an unnecessary "i" in i[VirtualCell].

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

**Comment by chokobole** [comment]
Sorry for ugly rebase..

File: tachyon/zk/plonk/error.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  kInvalidInstances,
+  // The constraint system is not satisfied.
+  kConstraintSystemFailure,
+  // Out of bounds index passed to a backend
+  kBoundsFailure,
+  // Opening error
+  kOpening,
+  // Transcript error
+  kTranscript,
+  // |k| is too small for the given circuit.
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/zk/plonk/error.h:28]: Oh it comes from rust comment and it looks like below.

```rust
    /// `k` is too small for the given circuit.
    NotEnoughRowsAvailable {
        /// The current value of `k` being used.
        current_k: u32,
    },
```

In this case it should be something like "The degree is too small for the given circuit."

File: tachyon/zk/plonk/circuit/assignment.h [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+class Assignment {
+ public:
+  using AnnotationCallback = base::OnceCallback<std::string()>;
+  using AssignCallback = base::OnceCallback<math::RationalField<F>()>;
+
+  // Creates a new region and enters into it.
+  //
+  // Panics if we are currently in a region (if |ExitRegion()| was not called).
+  //
+  // Not intended for downstream consumption; use |Layouter::AssignRegion()|
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/assignment.h:34]: Do you know where |Layouter::AssignRegion()| is coming from? I can't find it from both tachyon and halo2 implementation.

File: tachyon/zk/plonk/circuit/assignment.h [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  // Returns |Error::kChallengeUnavailable| if the current synthesis phase is
+  // before the challenge can be queried. Otherwise, it populates |value| with
+  // the challenge.
+  virtual Error GetChallenge(const Challenge& challenge, Value<F>* value) {
+    return Error::kChallengeUnavailable;
+  }
+
+  // Creates a new (sub)namespace and enters into it.
+  //
+  // Not intended for downstream consumption; use |Layouter::namespace()|
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/assignment.h:106]: Do you know where |Layouter::namespace()| is coming from? I can't find it from both tachyon and halo2 implementation.

File: tachyon/zk/plonk/circuit/table_column.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#define TACHYON_ZK_PLONK_CIRCUIT_TABLE_COLUMN_H_
+
+#include "tachyon/export.h"
+#include "tachyon/zk/plonk/circuit/column.h"
+
+namespace tachyon::zk {
+
+// A fixed column of a lookup table.
+//
+// A lookup table can be loaded into this column via |Layouter::AssignTable()].
```
Comment (TomTaehoonKim) [inline tachyon/zk/plonk/circuit/table_column.h:17]: Do you know where |Layouter::AssignRegion()| is coming from?

File: tachyon/zk/plonk/circuit/column_data.h [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+#include "absl/strings/substitute.h"
+
+#include "tachyon/export.h"
+#include "tachyon/zk/plonk/circuit/column_type.h"
+
+namespace tachyon::zk {
+
+struct TACHYON_EXPORT ColumnData {
+  ColumnType type = ColumnType::kAny;
+  size_t index = 0;
```
Comment (chokobole) [inline tachyon/zk/plonk/circuit/column_data.h:25]: Make getter

File: tachyon/zk/plonk/lookup/lookup_argument.h [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  size_t RequiredDegree() const {
+    CHECK_EQ(input_expressions_->size(), table_expressions_->size());
+
+    size_t max_input_degree = std::accumulate(
+        input_expressions_.begin(), input_expressions_.end(), 1,
+        [](size_t degree, const std::unique_ptr<Expression<F>>& input_expr) {
+          return std::max(degree, input_expr->Degree());
+        });
```
Comment (chokobole) [inline tachyon/zk/plonk/lookup/lookup_argument.h:60]: Use `max_element()`

File: tachyon/zk/plonk/constraint_system.h [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+    // TODO(chokobole): enable this.
+    // size_t degree = permutation_.RequiredDegree();
+
+    // The lookup argument also serves alongside the gates and must be accounted
+    // for.
+    size_t degree = std::max(degree, ComputeLookupRequiredDegree());
+
+    // Account for each gate to ensure our quotient polynomial is the
+    // correct degree and that our extended domain is the right size.
+    degree = std::cmp::max(degree, ComputeGateRequiredDegree());
```
Comment (chokobole) [inline tachyon/zk/plonk/constraint_system.h:366]:
```suggestion
    degree = std::max(degree, ComputeGateRequiredDegree());
```

File: tachyon/zk/plonk/constraint_system.h [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+    }
+    NOTREACHED();
+    return 0;
+  }
+
+  // Creates a new gate.
+  //
+  // A gate is required to contain polynomial constraints. This method will
+  // panic if |constrain| returns an empty iterator.
+  void CreateGate(const std::string_view& name, ConstrainCallback constrain) {
```
[the remark this answers is not in the document]
Comment (chokobole) [reply tachyon/zk/plonk/constraint_system.h:245]: I did it on purpose. for example, `annotate`.

**Review by TomTaehoonKim — APPROVED** [review]
LGTM! Great work!

**Review by Insun35 — APPROVED** [review]
LGTM

**Pull request merged**